documentation: remove unprefix hack after Pervasives deprecation

Octachron opened the pull request:
This PR build upon #1605 (in other words, only the two last commits are really part of this PR) and removes the unprefix hack currently used to build the documentation by adding some ad hoc logic to ocamldoc. I still need to investigate a handful of strange warnings but the manual and the stdlib documentation build correctly.

Octachron committed:
doc: remove unprefix trick

Octachron committed:
ocamldoc: adhoc support for prefixed stdlib

ghost commented:
It's nice that we can get rid of this hack! This PR can be rebased now that #1605 was merged.

Octachron commented:
I rebased the PR to make it much more readable; I still need to track four strange missing cross-references.

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -165,7 +165,7 @@ end = struct
   type t = Leaf of string | Node of ASet.t
   let compare t1 t2 =
     match (t1, t2) with
-    | (Leaf s1, Leaf s2) -> Pervasives.compare s1 s2
```
I was working on the manual and trunk broke the build, so I fixed this independently in c578cee935b5b0ecaac8eb3abc1d27fadc79de8c ( plus other minor fixes in dcbf1a8fae9bd313eabd8b9a8ec6511a0ef9ad22 ).

Octachron replied:
Thanks for the cleanup! I rebased this PR and fixed the testsuite in 2ac9f5ff6e5dd2e1 .

Octachron commented:
I found and fixed the last cross-reference issue and hide the prefix hack behing a `-lib` flag.

diml commented on manual/manual/Makefile:
```diff
@@ -47,13 +47,17 @@ html: htmlman/libref/style.css etex-files
 	mkdir -p htmlman/libref
-	$(OCAMLDOC) -colorize-code -sort -html \
+	cp $(SRC)/stdlib/doc_for_pervasives $(SRC)/stdlib/pervasives.mli
+	$(OCAMLDOC)-colorize-code -sort -html \
```
Nitpicking: this looked better with a space between `$(OCAMLDOC)` and `-colorize-code`. Without the space, it looks like the command is something like `ocamldoc-colorize-code`

Octachron replied:
Agreed.

diml commented on stdlib/doc_for_pervasives:
```diff
@@ -0,0 +1,2 @@
+[@@@deprecated "Use Stdlib"]
+include module type of struct include Stdlib end
```
A couple of thoughts:

- moving forward, we probably won't keep pervasive.ml up to date, i.e. new functions will only go in stdlib.mli. That means that the manual will show a few functions that don't exist
- transiently creating pervasives.mli could create a race condition in the build

What about systematically creating pervasives.mli with something like this?

```makefile
pervasives.mli: pervasives.ml
        echo '[@@@deprecated "Use Stdlib instead."]' > $@
        $(CAMLC) -nostdlib -I . -i $< >> $@
```

Octachron replied:
Creating pervasives.mli sounds like a better idea indeed.

Octachron replied:
I thought of an easier solution: just use `pervasives.ml` directly after adding a deprecated doc comments.

diml commented on stdlib/expand_module_aliases.awk:
```diff
@@ -22,8 +22,9 @@ NR == 1 { printf ("# 1 \"%s\"\n", FILENAME) }
   else if ($1 == "module")
-    printf ("\n(** @canonical %s *)\nmodule %s = Stdlib__%s%s\n",
-            $2, $2, tolower(substr($4,1,1)), substr($4,2));
+#   printf("\n(** @canonical %s *)", $2);
```
Why was this commented BTW? I believe it matters for odoc, for instance if someone tries to generate the HTML doc for all installed packages, it might look better with this tag

Octachron replied:
In this case, I fixed the awk script by not printing this line only if the variable ocamldoc is set to true. Like this, ocamldoc does not see an unknown custom tag, and odoc can generate better documentation.

diml replied:
Cool, thanks

Octachron committed:
ocamldoc: add a -lib flag for prefixed libraries

Octachron commented:
I have rebased and cleaned the PR's commits. A detail that I forgot to mention is that all stdlib modules now appears twice:, once as [Stdlib.Option](https://www.polychoron.fr/ocaml-nonmanual/reprefix/Stdlib.Option.html) and another time as [Option](https://www.polychoron.fr/ocaml-nonmanual/reprefix/Option.html) but they are not duplicated in the [index](https://www.polychoron.fr/ocaml-nonmanual/reprefix/index.html).

ghost commented:
Ok. It would be nicer if there was no duplication, but the current state of this PR is already an improvement over master, so we can leave the de-duplication as a future improvement.

gasche commented:
This commit breaks the CI under Windows(msvc):

```
CAML_LD_LIBRARY_PATH="../otherlibs/win32unix;../otherlibs/str" ../boot/ocamlrun ./ocamldoc -man -d stdlib_man -nostdlib -I ../stdlib -I ../parsing -I ../utils -I ../typing -I ../bytecomp -I ../driver -I ../otherlibs/str -I ../otherlibs/win32unix -I ../otherlibs/graphics -I ../otherlibs/dynlink -I ../otherlibs/systhreads \
-hide Stdlib  -lib Stdlib \
-pp "awk -v ocamldoc=true -f ../stdlib/expand_module_aliases.awk" \
-t "OCaml library" -man-mini ../stdlib/pervasives.ml ../stdlib/spacetime.mli ../stdlib/arg.mli ../stdlib/array.mli ../stdlib/arrayLabels.mli ../stdlib/bigarray.mli ../stdlib/buffer.mli ../stdlib/bytes.mli ../stdlib/bytesLabels.mli ../stdlib/callback.mli ../stdlib/camlinternalFormat.mli ../stdlib/camlinternalFormatBasics.mli ../stdlib/camlinternalLazy.mli ../stdlib/camlinternalMod.mli ../stdlib/camlinternalOO.mli ../stdlib/char.mli ../stdlib/complex.mli ../stdlib/digest.mli ../stdlib/ephemeron.mli ../stdlib/filename.mli ../stdlib/float.mli ../stdlib/format.mli ../stdlib/gc.mli ../stdlib/genlex.mli ../stdlib/hashtbl.mli ../stdlib/int32.mli ../stdlib/int64.mli ../stdlib/lazy.mli ../stdlib/lexing.mli ../stdlib/list.mli ../stdlib/listLabels.mli ../stdlib/map.mli ../stdlib/marshal.mli ../stdlib/moreLabels.mli ../stdlib/nativeint.mli ../stdlib/obj.mli ../stdlib/oo.mli ../stdlib/option.mli ../stdlib/parsing.mli ../stdlib/printexc.mli ../stdlib/printf.mli ../stdlib/queue.mli ../stdlib/random.mli ../stdlib/result.mli ../stdlib/scanf.mli ../stdlib/seq.mli ../stdlib/set.mli ../stdlib/stack.mli ../stdlib/stdLabels.mli ../stdlib/stdlib.mli ../stdlib/stream.mli ../stdlib/string.mli ../stdlib/stringLabels.mli ../stdlib/sys.mli ../stdlib/uchar.mli ../stdlib/weak.mli ../parsing/ast_invariants.mli ../parsing/parsetree.mli ../parsing/printast.mli ../parsing/depend.mli ../parsing/docstrings.mli ../parsing/parser.mli ../parsing/longident.mli ../parsing/lexer.mli ../parsing/asttypes.mli ../parsing/attr_helper.mli ../parsing/syntaxerr.mli ../parsing/location.mli ../parsing/parse.mli ../parsing/pprintast.mli ../parsing/builtin_attributes.mli ../parsing/ast_helper.mli ../parsing/ast_mapper.mli ../parsing/camlinternalMenhirLib.mli ../parsing/ast_iterator.mli ../utils/config.mli ../utils/build_path_prefix_map.mli ../utils/identifiable.mli ../utils/consistbl.mli ../utils/warnings.mli ../utils/numbers.mli ../utils/misc.mli ../utils/profile.mli ../utils/strongly_connected_components.mli ../utils/arg_helper.mli ../utils/ccomp.mli ../utils/targetint.mli ../utils/clflags.mli ../utils/terminfo.mli ../typing/typemod.mli ../bytecomp/simplif.mli ../driver/pparse.mli ../otherlibs/str/str.mli ../otherlibs/unix/unix.mli ../otherlibs/unix/unixLabels.mli ../otherlibs/systhreads/thread.mli ../otherlibs/systhreads/condition.mli ../otherlibs/systhreads/mutex.mli ../otherlibs/systhreads/event.mli ../otherlibs/systhreads/threadUnix.mli ../otherlibs/graph/graphics.mli ../otherlibs/graph/graphicsX11.mli ../otherlibs/dynlink/dynlink.mli
'awk' n’est pas reconnu en tant que commande interne
ou externe, un programme exécutable ou un fichier de commandes.
Preprocessing error
Error while running external preprocessor
Command line: awk -v ocamldoc=true -f ../stdlib/expand_module_aliases.awk "../stdlib/pervasives.ml" > C:\cygwin64\tmp\ocamlpp34f5d6
```

gasche commented:
I believe that we have an AWK variable that points to `awk` or `gawk` as appropriate.

Octachron commented:
Indeed, there is AWK, I am waiting for CI then I will commit this [fix](https://github.com/Octachron/ocaml/commit/284c8f8bb858009ebba5ab34b3a8ec493bb8f7bf).

gasche commented:
Note: other Makefiles seem to use `awk` just fine (the `report` target in testsuite/Makefile) and I don't understand why... It might be related to the context in which the command is run (from Make directly, or invoked by ocaml in a `-pp` command), but always using $AWK would be the better behavior I think.

nojb commented:
I don't think this is going to work. `awk`/`gawk` are Cygwin programs, but the the `-pp` argument is invoked by the compiler, so is executed with `cmd.exe`.

gasche commented:
*Someone* may have to rewrite the script as an OCaml program...

Octachron commented:
I will rewrite the script as an OCaml program if it is needed, but `stdlib/Compflags`  already uses 
` ' -pp "$AWK -f expand_module_aliases.awk"'` ? ( I am waiting for the right to run precheck on the INRIA CI)

nojb commented:
Good point. Indeed, it should work because `cmd.exe` inherits the PATH from Cygwin.

nojb commented:
FYI, I just built `trunk` on Windows with the above fix and it works.

Octachron commented:
@nojb , thanks for the cross-checking! Since precheck was also green, I pushed 284c8f8bb858009ebba5ab34b3a8ec493bb8f7bf .

Sorry for the CI breakage.

gasche commented:
> Sorry for the CI breakage.

No problem, this is the very point of having a CI in the first place :-)
